src/network/ledger.rs

```diff
@@ -812,16 +812,35 @@ impl<N: Network, E: Environment> Ledger<N, E> {
                 return;
             }
 
+            // Filter out any pre-existing block requests for the peer.
+            let mut new_block_heights = Vec::new();
+            if let Some(block_requests) = self.block_requests.read().await.get(&peer_ip) {
+                for block_height in start_block_height..=end_block_height {
+                    if !block_requests.contains_key(&block_height.into()) {
+                        new_block_heights.push(block_height);
+                    }
+                }
+            } else {
+                self.add_failure(peer_ip, format!("Missing block requests for {}", peer_ip)).await;
```

Comment:
You might want to add a `return;` here to match the original logic, otherwise it would technically still attempt to acquire a `write` lock in the next step.